src/relay/dnsrelay/mod.rs

```diff
@@ -123,65 +122,70 @@ async fn acl_lookup(
     remote: &Address,
     qname: &Name,
     qtype: RecordType,
-) -> io::Result<Message> {
+) -> io::Result<(Message, bool)> {
     // Start querying name servers
     debug!(
         "attempting lookup of {:?} {} with ns {:?} and {:?}",
         qtype, qname, local, remote
     );
 
-    let timeout = Some(Duration::new(3, 0));
-    let local_response_fut = try_timeout(local_lookup(qname, qtype, local), timeout);
-
-    let timeout = Some(Duration::new(3, 0));
-    let remote_response_fut = try_timeout(proxy_lookup(context.clone(), svr_cfg, remote, qname, qtype), timeout);
-
-    let (local_response, remote_response) = future::join(local_response_fut, remote_response_fut).await;
-    let local_response = local_response.unwrap_or_else(|_| Message::new());
-    let remote_response = remote_response.unwrap_or_else(|_| Message::new());
-
     // remove the last dot from fqdn name
     let mut name = qname.to_ascii();
     name.pop();
     let addr = Address::DomainNameAddress(name, 0);
-    let qname_in_proxy_list = context.check_qname_in_proxy_list(&addr).await;
-
-    let mut ip_bypassed = false;
-    for rec in local_response.answers() {
-        let bypassed = match rec.rdata() {
-            RData::A(ref ip) => {
-                let addr = Address::SocketAddress(SocketAddr::new(IpAddr::from(*ip), 0));
-                context.check_target_bypassed(&addr).await
-            }
-            RData::AAAA(ref ip) => {
-                let addr = Address::SocketAddress(SocketAddr::new(IpAddr::from(*ip), 0));
-                context.check_target_bypassed(&addr).await
+    let qname_in_proxy_list = context.check_qname_in_proxy_list(&addr);
+
+    let remote_response_fut = async {
```

Comment:
I forgot to invoke `tokio::spawn` here. Is there any way to do this without inappropriate lifetimes?

Reply:
All the scoped libraries I found seems to swallow return values of the future, which is undesirable...

Reply:
I think previous join approach is better?

Reply:
We do not want to send out both probes all the time though. If the hostname is matched, certainly we only need to send one probe. This improves privacy slightly (use case: you want to hide even DNS traffic to some sensitive website).

Reply:
This PR looks promising, alas it is a WIP: https://github.com/tokio-rs/tokio/pull/2153/files#diff-d0c8061133c518b906f05037bd563ed7R127

Reply:
Hmm, but if we spawn here, actually we send the remote request ahead as well?

Reply:
@madeye Look at the next two lines below.

Reply:
What about this?